database/migrations/000013_rule_evaluations_rule_name.up.sql

```diff
@@ -0,0 +1,32 @@
+-- Copyright 2024 Stacklok, Inc
+--
+-- Licensed under the Apache License, Version 2.0 (the "License");
+-- you may not use this file except in compliance with the License.
+-- You may obtain a copy of the License at
+--
+--      http://www.apache.org/licenses/LICENSE-2.0
+--
+-- Unless required by applicable law or agreed to in writing, software
+-- distributed under the License is distributed on an "AS IS" BASIS,
+-- WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+-- See the License for the specific language governing permissions and
+-- limitations under the License.
+
+-- Begin transaction
+BEGIN;
+
+-- Add rule_name to rule_evaluations
+ALTER TABLE rule_evaluations
+  ADD COLUMN rule_name TEXT;
+
+-- Drop the existing unique index on rule_evaluations
+DROP INDEX IF EXISTS rule_evaluations_results_idx;
+
+-- Recreate the unique index with rule_name
+CREATE UNIQUE INDEX rule_evaluations_results_idx
+  ON rule_evaluations (profile_id, repository_id, COALESCE(artifact_id, '00000000-0000-0000-0000-000000000000'::UUID),
+                       entity, rule_type_id, COALESCE(pull_request_id, '00000000-0000-0000-0000-000000000000'::UUID),
+                       rule_name) NULLS NOT DISTINCT;
```

Comment:
Correct. The reason we use `NULLS NOT DISTINCT` is because _if we don't_, two rule evaluations, which are identical would now be unique and we would have something like (omitted some cols):

| entity | profile_id | rule_type_id | rule_name | 
|--------|------------|--------------|-----------|
|   repo     |     1       |     1         |   null        |   
|   repo     |     1       |     1         |   null        |   

Now this is good for the meantime, we would also fix evaluations for existing profiles, but when we try to backfill, we can end up with:

| entity | profile_id | rule_type_id | rule_name | 
|--------|------------|--------------|-----------|
|   repo     |     1       |     1         |   abc_rule        |   
|   repo     |     1       |     1         |   abc_rule        |   

Leading to a violation of our index. That's why we require `NULLS NOT DISTINCT` .

> And since we're in a transaction, dropping the old index before adding the new is also safe; the table should only be locked briefly at our current data size.

Yes, it is safe.

Reply:
To be double sure, I checked the docs again, and:

From https://www.postgresql.org/docs/15/sql-dropindex.html
> A normal DROP INDEX acquires an ACCESS EXCLUSIVE lock on the table, blocking other accesses until the index drop can be completed. 

From https://www.postgresql.org/docs/15/explicit-locking.html
> Once acquired, a lock is normally held until the end of the transaction.

We don't have any savepoints so lock won't be released until the end of the transaction.